### PR TITLE
Initial authorization implementation

### PR DESCRIPTION
**Pull request by Janpot**
Getting started on the infrastructure needed to support per-page authorization

- [x] Add a new file `application.yml` at the root
- [x] Allow defining roles in that file under `authorization.roles`. These are the roles available in the Toolpad application
- [x] Allow defining roles on the page level. These are used to regulate access to the page
- [x] Page roles editor
- [x] Application roles editor
- [x] Block pages depending on the role of authenticated user (depends on [authentication](https://github.com/mui/mui-toolpad/pull/2925))
- [ ] documentation

### CODE REVIEW

File: packages/toolpad-app/src/toolpad/AppEditor/PageEditor/PageOptionsPanel.tsx [hunk 1]
```diff
@@ -37,52 +44,111 @@ export default function PageOptionsPanel() {
...
+            control={
+              <Checkbox checked={!page.attributes.authorization} onChange={handleAllowAllChange} />
+            }
+            label="Allow access to all roles"
+          />
+          <Autocomplete
+            multiple
+            options={Array.from(availableRoles.keys())}
+            value={page.attributes.authorization?.allowedRoles ?? []}
+            onChange={handleallowedRolesChange}
```
Comment (apedroferreira) [inline packages/toolpad-app/src/toolpad/AppEditor/PageEditor/PageOptionsPanel.tsx:135]:
```suggestion
            onChange={handleAllowedRolesChange}
```

File: packages/toolpad-app/src/toolpad/AppEditor/PageEditor/PageOptionsPanel.tsx [hunk 2]
```diff
@@ -37,52 +44,111 @@ export default function PageOptionsPanel() {
     [domApi, page],
   );
 
+  const availableRoles = React.useMemo(() => {
+    return new Map(appNode.attributes?.authorization?.roles?.map((role) => [role.name, role]));
+  }, [appNode]);
+
+  const handleallowedRolesChange = React.useCallback(
```
Comment (apedroferreira) [inline packages/toolpad-app/src/toolpad/AppEditor/PageEditor/PageOptionsPanel.tsx:51]:
```suggestion
  const handleAllowedRolesChange = React.useCallback(
```

**Review by apedroferreira — APPROVED** [review]
This looks good - should be pretty straightforward to integrate with authentication, I'll so that next.

In terms of UX, I guess the position of button to open the authorization model is just temporary? We can find a better place for it later.
And looks like the only way to save a role is to press "Enter"? I didn't realize that at first, filled in name and description and only then realized that my role was not saved.

**Comment by apedroferreira** [comment]
Also just noticed this seems to be out of place?

Screenshot 2023-12-06 at 18 34 33

**Comment by Janpot** [comment]
> And looks like the only way to save a role is to press "Enter"? I didn't realize that at first, filled in name and description, closed the modal and only then realized that my role was not saved.

Can't seem to reproduce this:

https://github.com/mui/mui-toolpad/assets/2109932/e8627468-9cf8-437e-aea8-a60c260a74ba

**Comment by apedroferreira** [comment]
> Can't seem to reproduce this:

Weird, now I can't either... Nevermind I guess.

**Pull request merged**